Sming/Arch/Esp32/Core/HardwarePWM.cpp

```diff
@@ -140,13 +140,14 @@ uint32_t maxDuty(ledc_timer_bit_t bits)
 
 HardwarePWM::HardwarePWM(uint8_t* pins, uint8_t no_of_pins) : channel_count(no_of_pins)
 {
-	debug_d("starting HardwarePWM init");
+	assert(no_of_pins > 0 && no_of_pins <= SOC_LEDC_CHANNEL_NUM);
+	no_of_pins = std::min(uint8_t(SOC_LEDC_CHANNEL_NUM), no_of_pins);
+	debug_i("initializing PERIPH_LEDC_MODULE");
 	periph_module_enable(PERIPH_LEDC_MODULE);
-	if((no_of_pins == 0) || (no_of_pins > SOC_LEDC_CHANNEL_NUM)) {
-		return;
-	}
 
 	for(uint8_t i = 0; i < no_of_pins; i++) {
+		//make sure we don't try to use pins unavailable for the SoC
+		assert(SOC_GPIO_VALID_OUTPUT_GPIO_MASK & (1U << pins[i]));
```

Comment:
If you look at `ledc_channel_config` you'll see that this exact check is already provided `LEDC_ARG_CHECK(GPIO_IS_VALID_OUTPUT_GPIO(gpio_num), "gpio_num");`

Reply:
sigh. I came from observing the bug that I described above (my current setting, as used on the 8266,  uses the two flash pins). First, I started with asserts that would have tripped on using those, but then thought that this is the wrong place to do this.
I'm now adding a list of allowable pins per SoC to the configuration (that would be a great use case for defaults for arrays with objects in ConfigDB, btw).
This assert just seemed to be the minimal sensible thing to do. I should have assumed that someone was already sensible enough to add it.
I'll close the pr then.

Reply:
Yeah, I've seen designs for the esp8266 which doesn't use flash at all but boots over SDIO so can't even assume they'll never be used!